src/python/torchdistx/slowmo/slowmo_optimizer.py

```diff
@@ -0,0 +1,221 @@
+        >>>  slowmo_state = slowmo_comm.SlowMoState(
+        >>>     cur_subgroup,
+        >>>     sync_grads=True
+        >>>  )
+        >>>
+        >>>  # Register SlowMo hook, which only communicates gradients
+        >>>  # in a intra-node fashion.
+        >>>  fsdp_net.register_comm_hook(
+        >>>     slowmo_state,
+        >>>     slowmo_comm.slowmo_hook
+        >>>  )
+        >>>
+        >>>  base_optimizer = torch.optim.SGD(
+        >>>     fsdp_net_slowmo.parameters(),
+        >>>     lr=1e-2
+        >>>  )
+        >>>  # Create a SlowMo optimizer that wraps a local optimizer.
+        >>>  slowmo_optim = slowmo_optimizer.SlowMomentumOptimizer(
+        >>>     base_optim=base_optimizer,
+        >>>     slowmo_freq=6,
+        >>>     slowmo_factor=0.5,
+        >>>     slowmo_lr=0.1
+        >>>  )
+        >>>
+        >>>  # SlowMo runs intra-node gradient averaging at every step,
+        >>>  # every 6th step it will run model averaging and
+        >>>  # a slow momentum update.
+        >>>  for step in range(200):
+        >>>     slowmo_optim.zero_grad()
+        >>>     loss = loss_fn(output, labels)
+        >>>     loss.backward()
+        >>>     slowmo_optim.step()
+    """
+
+    def __init__(
+        self,
+        base_optim: torch.optim.Optimizer,
+        slowmo_freq: int = 48,
+        slowmo_factor: float = 0.5,
+        slowmo_lr: float = 1.0,
+    ):
+        if base_optim is None:
+            raise ValueError("Base optimizer is a required parameter.")
+        self._base_optim = base_optim
+
+        # check that base optimizer's learning rate is stored in param_groups
+        if not (self._base_optim.param_groups):
+            raise ValueError(
+                "Provided base optimizer does not have parameters specified."
+            )
+        self.param_groups = self._base_optim.param_groups
+
+        if slowmo_freq < 1:
+            raise ValueError(
+                "Invalid ``slowmo_freq`` parameter, must be a positive value."
+            )
+        self.slowmo_freq = slowmo_freq
+
+        if slowmo_factor < 0.0:
+            raise ValueError(
+                "Invalid ``slowmo_factor`` parameter, must be non-negative."
+            )
+        self.slowmo_factor = slowmo_factor
+
+        if slowmo_lr < 0.0:
+            raise ValueError("Invalid ``slowmo_lr`` parameter, must be non-negative.")
+        self.slowmo_lr = slowmo_lr
+
+        self.averager = averagers.PeriodicModelAverager(
+            period=slowmo_freq, warmup_steps=0
+        )
+        self.buffers_initialized = False
+
+        # Memorize initial parameters. Can't put them in the `state`,
+        # because some of optimizers rely on the state being empty during the `step()`
+        # to put specific information into the state. The very first set of parameters
+        # should be remembered before the first `step()`,
+        # thus can't put this info into `state`.
```

Comment:
nit: Removing some verbosity
```suggestion
        # Memorize initial parameters before the first `step()`. Can't put them in `self.state`,
        # because some of optimizers rely `self.state` being empty during the `step()`
        # to initialize optimizer states.
        # `self._prev_parameters` must be in sync with the flattened version of `self.param_groups`.
```
Plus, we should document that invariant about `self._prev_parameters` since the internal implementation requires it for correctness.